File: rust/src/leetcode/Q152_maximum_product_subarray.rs
/*
 * @lc app=leetcode id=152 lang=rust
 *
 * [152] Maximum Product Subarray
 */

// @lc code=start
impl Solution {
    pub fn max_product(nums: Vec<i32>) -> i32 {
        let mut result: i32 = nums[0];
        let mut dp_min: Vec<i32> = vec![nums[0]; nums.len()];
        let mut dp_max: Vec<i32> = vec![nums[0]; nums.len()];

        for i in 1..nums.len() {
            let num: i32 = nums[i];

            if num >= 0 {
                dp_min[i] = num.min(num * dp_min[i - 1]);
                dp_max[i] = num.max(num * dp_max[i - 1]);
            } else {
                dp_min[i] = num.min(num * dp_max[i - 1]);
                dp_max[i] = num.max(num * dp_min[i - 1]);
            }

            result = result.max(dp_max[i]);
        }

        result
    }
}
// @lc code=end

File: python/Q152_maximum_product_subarray.py
#
# @lc app=leetcode id=152 lang=python3
#
# [152] Maximum Product Subarray
#


# @lc code=start
class Solution:
    def maxProduct(self, nums: list[int]) -> int:
        min_dp: list[int] = [nums[0]] * len(nums)
        max_dp: list[int] = [nums[0]] * len(nums)
        result: int = nums[0]

        for i in range(1, len(nums)):
            number: int = nums[i]

            if number > 0:
                min_dp[i] = min(number, number * min_dp[i - 1])
                max_dp[i] = max(number, number * max_dp[i - 1])
            elif number < 0:
                min_dp[i] = min(number, number * max_dp[i - 1])
                max_dp[i] = max(number, number * min_dp[i - 1])
            else:
                min_dp[i] = 0
                max_dp[i] = 0

            if max_dp[i] > result:
                result = max_dp[i]

        return result


# @lc code=end
